### PR TITLE
Added keyword highlighting

### PR DESCRIPTION
**Pull request by benfrancis314**
Added functionality for keyword highlighting. This was branched from my ben_temp2 branch before it was sent for the pull request yesterday, so some of the problems with that pull request are still here (the pull request that added the display). When I merge this PR into master after we merge the display PR into master I'll make sure only the up to date changes are included.

### CODE REVIEW

**Comment by benfrancis314** [comment]
Hmmm in handling the conflicts there was one occasion where I selected this branch's changes even though the master was better, for the reason that I think they would have made the master unstable (it was just a refactoring). I figured I'd prefer a stable master and just go back later and fix it quickly in a separate PR, which would have been necessary either way

File: extension/content_script.js [hunk 1]
```diff
@@ -11,8 +11,9 @@ if (window[namespace] === true) {
 
 // Keeps track of the pointed text. Initialized by end of script load.
 var tracker = null;
-// Creates display for time remaining and reading speed at top of page. Initialized by end of script load.
```
Comment (sjonany) [inline extension/content_script.js:14]: comment got deleted?

File: extension/content_script.js [hunk 2]
```diff
@@ -251,11 +256,21 @@ function parseDocument() {
 }
 
 let readableDomIds = parseDocument();
+doc = new Doc(readableDomIds);
 tracker = new Tracker(readableDomIds);
-display = new Display(readableDomIds, speed);
+window.tracker = tracker;
```
Comment (sjonany) [inline extension/content_script.js]: is this global variable used?

File: extension/content_script.js [hunk 3]
```diff
@@ -251,11 +256,21 @@ function parseDocument() {
 }
 
 let readableDomIds = parseDocument();
+doc = new Doc(readableDomIds);
 tracker = new Tracker(readableDomIds);
-display = new Display(readableDomIds, speed);
+window.tracker = tracker; 
+display = new Display(readableDomIds, speed, doc.getTotalWords());
+window.display = display;
```
Comment (sjonany) [inline extension/content_script.js]: hm did we revert the change where this is no longer a global var?

File: extension/lib/tracker.js [hunk 4]
```diff
@@ -106,6 +106,10 @@ class Tracker {
         return $("#" + this.readableDomIds[containerId]);
     }
 
+    getFirstContainer() { // Added to access container ID
```
Comment (sjonany) [inline extension/lib/tracker.js]: was this file reverted?
i think we agreed that
- getFirstContainer() is not used
- window.display below should not be used cos we want classes like tracker to be self contained.

File: extension/content_script.js [hunk 5]
```diff
@@ -251,11 +256,21 @@ function parseDocument() {
...
+
+
+
 setupClickListener(tracker);
 readListener();
 
+startMove(direction.FORWARD); // Start reader on the first line
```
Comment (sjonany) [inline extension/content_script.js:302]: [just a comment, please resolve] --> in google we have a way of indicating these (comments that don't have to be resolved), but idk how to in github :(
ah, a behavior change. i agree w/ this.

File: extension/lib/display.js [hunk 6]
```diff
@@ -23,10 +23,10 @@ class Display {
     - readableDomIds: string[]. List of dom IDs that contain readable content.
     - speed: int. Value proportional to the speed of auto-read mode, before adjustment by sentence length. 
     */
-    constructor(readableDomIds, speed) {
+    constructor(readableDomIds, speed, total_words) {
```
Comment (sjonany) [inline extension/lib/display.js:27]: pls add comment. i guess this is the total number of tokens / words in the entire document?

File: extension/lib/display.js [hunk 7]
```diff
@@ -68,25 +68,17 @@ class Display {
     Return: 
     The total time remaining to read document, based on avg letters/word and lowball estimate of WPM (int)
     */
-    initTimer(readableDomIds) {
-        let total_char = 0; // NEED TO REFACTOR TOTAL_CHAR CALCULATION
-        let avg_letters_per_word = 6; // Actual is 4.79 (from Quora), add a space after each word to get 6. 
+    initTimer(readableDomIds, total_words) {
         let avg_read_speed = 200; // Low ball estimate, from Rayner (in WPM)
```
Comment (sjonany) [inline extension/lib/display.js:73]: been meaning to ask in the previous cl, so ill ask while i remember.
Q: why don't we estimate using the actual speed currently set by the user? wouldn't that would be the most faithful estimation?
-- If user has not changed speed at all, then shouldn this just be the default speed?

Comment (sjonany) [reply extension/lib/display.js:73]: If non trivial code, feel free to leave TODO for a future CL

File: extension/lib/doc.js [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+/*
+Class dedicated to keeping track of model of document. 
+Note: Called "Doc" because Document is already used in js. 
+*/
+
+class Doc {
+    constructor(readableDomIds) {
+        this.readableDomIds = readableDomIds; // string[]; List of all container IDs
+        this.containerId = null; // str; ID of current container
+        this.dict = null; // { str : int } , Set in calcTotalWords function
```
Comment (sjonany) [inline extension/lib/doc.js]: if too tiring to refactor, im fine if u leave it like this, but...
consider choosing var name that describes the fcn, rather than just the type
some choices:
- wordTally
- wordCount

if u were to also include the type as part of the name im cool w/ that too.
E.g.
- wordCountDict
But i think the function of the variable is more important.

File: extension/lib/doc.js [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        this.keywords = this.setKeyWords(this.dict); // string[] keywords of document
+    };
+
+    // Returns: Total words in doc (int)
+    getTotalWords() {
+        return this.total_words;
+    }
+
+    // Returns: Container given container ID (jQuery element)
+    getContainer(containerId) {
```
Comment (sjonany) [inline extension/lib/doc.js:36]: can you leave TODO: refactor?
i think this is redundant w/ tracker.js
i might create a domWrapper class that will be shared by both tracker and this guy, but not sure. would need to look at the final codebase after all the pulls

File: extension/lib/doc.js [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+            };
+        };
+        return keywords;
+    };
+
+    /*
+    Params: Current container, start and end of tracker (jQuery element, int, int)
+    Highlights the keywords within the tracked sentence. 
+    */
+    highlightKeyWords(container, start, end) {
```
Comment (sjonany) [inline extension/lib/doc.js]: this method i prefer to be in content_script.js
initially i was expecting doc.js to be "all the NLP logic. no UI prettiness"
but now we also add the UI drawing stuff in this class.
another downside is when someone wants to do ui stuff they'll collide w/ someone who's trying to do more nlp wizardry.

Can leave a TODO for future work, but might be something like this:
- doc.js identifies the keyword offsets within the container and return them as a list
- content_script.js does the ui highlighting.

What do you think?

**Review by sjonany — CHANGES_REQUESTED** [review]

**Review by sjonany — COMMENTED** [review]
programming tip. if u see lots of [wordlList[i]], might be worth refactoring to let word = wordList[i].
pretty trivial here, so doesn't really matter, but if u get used to this, u might not only make code more readable, but also more performant.
e.g. u save some lookup cost of [i].

File: extension/content_script.js (hunk 1 above)
Comment (benfrancis314) [reply extension/content_script.js:14]: Hmm *look of confusion, adds back in comment*

File: extension/content_script.js (hunk 2 above)
Comment (benfrancis314) [reply extension/content_script.js]: No but it was in the pull request before, this was kinda weird bc I branched this from the PR that you just reviewed before the comments on it, instead of master bc master was a bit behind at the time ... will go back and fix it tho

File: extension/content_script.js (hunk 3 above)
Comment (benfrancis314) [reply extension/content_script.js]: I believe it is now in master, but this is a relic of a different branch before we fixed it

File: extension/lib/tracker.js (hunk 4 above)
Comment (benfrancis314) [reply extension/lib/tracker.js]: Yep; see above comments (this branched from the most recent pull request pre-review, when these problems were still in there). I was hoping they'd be fixable in merge conflicts, it appears this escaped, will fix

File: extension/lib/display.js (hunk 7 above)
Comment (benfrancis314) [reply extension/lib/display.js:73]: So we have the auto-scroll start at 400 WPM (-ish; again it's just an estimate), but we calculate the reading time with a low-ball estimate of 200 WPM. 200 WPM is actually almost dreadfully slow (assuming they read at a constant speed, which the autoscroll does). I suspect the 200 WPM is more of a result of a distraction than reading speed, so I think it is valid to incorporate it in a realistic "total time til yer done", but the auto-read should be faster than that, since that is supposed to measure literally how many words go by per second, not accounting for things like distractions, and it is only useful if it is around 400 WPM

File: extension/lib/doc.js (hunk 8 above)
Comment (benfrancis314) [reply extension/lib/doc.js]: Oh oops I was literally thinking of a dictionary in the physical sense (like a book that contains words), I agree this should for sure be changed

Comment (benfrancis314) [reply extension/lib/doc.js]: Either way a poor name

File: extension/lib/doc.js (hunk 10 above)
Comment (benfrancis314) [reply extension/lib/doc.js]: Yep I agree with internal logic 100%, was mostly a time consideration. Actually ended up not being hard, will include in next pr

Comment (benfrancis314) [reply extension/lib/doc.js]: I actually was able to just move the entire func into contentScript, which just gets the keywords from the doc class. Does that jive with what you were thinking?

File: extension/lib/display.js (hunk 7 above)
Comment (benfrancis314) [reply extension/lib/display.js:73]: But good point that if a user is using the autoscroll we will want to adjust the estimated time

File: extension/content_script.js [hunk 11]
```diff
@@ -149,18 +151,55 @@ function scroll() {
...
+	// Get list of words in interval
+	let containerText = container.text();
+	let sentenceText = containerText.slice(start,end);
+	let wordRegex = /\b\w+\b/g;
+	let wordList = sentenceText.match(wordRegex);
+	let keywords = doc.getKeyWords();
+	for (var i in wordList) { // Accentuate keywords
+		// TODO: This only gets the first occurence of each word in the sentence; should get all
+		let word = wordList[i];
+		if (keywords.includes(word.toLowerCase())) { // See if each word is a keyword
```
Comment (sjonany) [inline extension/content_script.js]: prefer to use set() instead of array when you only want to check if the data structure contains an element
- https://alligator.io/js/sets-introduction/

array's hasElement() is O(n)
hashset's hasElement() is O(1) on average
--- If the above syntax isn't familiar, lmk and i can find useful resources.

File: extension/lib/display.js (hunk 7 above)
Comment (sjonany) [reply extension/lib/display.js:73]: Q1. Do we agree that "if in autoscroll, when user increase his wpm, then the estimated time should decrease?"
This means that we want estimated time during autoscroll = words left / current autoscroll wpm

Q2. If Q1 is a yes, then _outside_ of autoscroll, what wpm should we use for estimated time?
I just assumed it should be autoscroll wpm. If we use other wpm's, then as soon as user presses alt, the estimated time would suddenly decrease by a lot, which might be confusing to the user.

And... maybe your comment means that the default autoscroll wpm should be 200 wpm. It was unbearable slow before the animation, but w/ the animation, maybe it's more tolerable?

File: extension/lib/doc.js (hunk 9 above)
Comment (sjonany) [reply extension/lib/doc.js:36]: Oh, please add the comment that this is redundant w/ tracker.js ' getContainer so I'll remember what the todo is about.

File: extension/lib/doc.js (hunk 10 above)
Comment (sjonany) [reply extension/lib/doc.js]: yeah!
i was thinking a slightly different version where u return the _indices_ of the interesting keywords. when later on u want to use fancier nlp-based logic u are completely shielding the ui-interested people from the nlp logic.
but im fine w/ this rn.

File: extension/lib/display.js (hunk 7 above)
Comment (benfrancis314) [reply extension/lib/display.js:73]: Ikk I am not sure about this predicament; perhaps more user testing would elucidate

Though I don't think we should move autoscroll down to 200 for the purpose of the timer, I think the current speed is actually pretty good. (My mom was actually significantly faster than it and immediately sought out how to make it faster)

File: extension/lib/doc.js (hunk 10 above)
Comment (benfrancis314) [reply extension/lib/doc.js]: Oo yes I like that for later as well

File: extension/content_script.js (hunk 11 above)
Comment (benfrancis314) [reply extension/content_script.js]: Oh yep, I thought about it for a second but was just trying to rapidly get to testing, definitely good to implement before master merge though

Also thanks for the js set link, I figured js had sets but was not directly familiar with

Comment (benfrancis314) [reply extension/content_script.js]: Hmm I'm going to go to bed right now, can I address this in a separate pull request in the morning so that you can merge yours in as well (if you're going to tonight; if you're going to tomorrow no worries)?

I'll repush everything but this, and if you want you can just merge this and I'll do another CR tmrw mrnng
OR if you don't then I'll just address this then repush, CR again
